packages/aws-cdk/lib/list-stacks.ts

```diff
@@ -0,0 +1,98 @@
+
+import '@jsii/check-node/run';
+import { Environment } from '@aws-cdk/cx-api';
+import { DefaultSelection, ExtendedStackSelection, StackCollection } from './api/cxapp/cloud-assembly';
+import { CdkToolkit } from './cdk-toolkit';
+
+/**
+ * Options for List Stacks
+ */
+export interface ListStacksOptions {
+  /**
+   * Stacks to list
+   *
+   * @default - All stacks are listed
+   */
+  readonly selectors: string[];
+}
+
+/**
+ * Type to store stack dependencies recursively
+ */
+export type DependencyDetails = {
+  id: string;
+  dependencies: DependencyDetails[];
+};
+
+/**
+ * Type to store stack and their dependencies
+ */
+export type StackDetails = {
+  id: string;
+  name: string;
+  environment: Environment;
+  dependencies: DependencyDetails[];
+};
+
+/**
+ * List Stacks
+ *
+ * @param toolkit cdk toolkit
+ * @param options list stacks options
+ * @returns StackDetails[]
+ */
+export async function listStacks(toolkit: CdkToolkit, options: ListStacksOptions): Promise<StackDetails[]> {
+  const assembly = await toolkit.assembly();
+
+  const stacks = await assembly.selectStacks({
+    patterns: options.selectors,
+  }, {
+    extend: ExtendedStackSelection.Upstream,
+    defaultBehavior: DefaultSelection.AllStacks,
+  });
+
+  function calculateStackDependencies(collectionOfStacks: StackCollection): StackDetails[] {
+    const allData: StackDetails[] = [];
+
+    for (const stack of collectionOfStacks.stackArtifacts) {
+      const data: StackDetails = {
+        id: stack.id,
+        name: stack.stackName,
+        environment: stack.environment,
+        dependencies: [],
+      };
+
+      for (const dependencyId of stack.dependencies.map(x => x.id)) {
+        if (dependencyId.includes('.assets')) {
+          continue;
+        }
+
+        const depStack = assembly.stackById(dependencyId);
+
+        if (depStack.stackArtifacts[0].dependencies.length > 0 &&
```

Comment:
Just curious -- why are we only getting the first `stackArtifact` here?

Reply:
So we are getting the dependent stack by ID: `assembly.stackById(dependencyId);`, so there must be just one unique stack by id. 

I can add a comment explaining this or/and initially I had an error if it was greater than one but I believed that validation should be done before this (I can check if that is present). Thougths?

Reply:
Ok, so [stackById](https://github.com/aws/aws-cdk/blob/7205143d563449ab211406da04d99bbbcdc7c8cc/packages/aws-cdk/lib/api/cxapp/cloud-assembly.ts#L206-L212) uses [getStackArtifact](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/cx-api/lib/cloud-assembly.ts#L110) which would always return one artifact but `stackById` wraps it in an array.

Reply:
Thanks for the clarification. Not now, unrelated, and also not sure what the impact would be, but maybe this is something we can update in a future PR. It feels like `stackById` should return a single stack, not an array containing a single stack.

Reply:
I agree :)